sycl/include/sycl/ext/oneapi/group_algorithm.hpp

```diff
@@ -65,6 +65,125 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
+  auto DestP =
+      multi_ptr<uint8_t, DestS>(reinterpret_cast<uint8_t *>(Dest.get()));
+  auto SrcP = multi_ptr<uint8_t, SrcS>(reinterpret_cast<uint8_t *>(Src.get()));
+  return async_group_copy(g, SrcP, DestP, NumElements, Stride);
+}
+
+/// Specialization for vector bool type.
+/// Asynchronously copies a number of elements specified by \p NumElements
+/// from the source pointed by \p Src to destination pointed by \p Dest
+/// with a stride specified by \p Stride, and returns a SYCL device_event
+/// which can be used to wait on the completion of the copy.
+template <typename Group, typename T, access::address_space DestS,
+          access::address_space SrcS>
+detail::enable_if_t<is_group_v<Group> && detail::is_vector_bool<T>::value,
+                    device_event>
+async_group_copy(Group g, multi_ptr<T, SrcS> Src, multi_ptr<T, DestS> Dest,
+                 size_t NumElements, size_t Stride) {
+  static_assert(sizeof(bool) == sizeof(uint8_t),
+                "Async copy to/from bool memory is not supported.");
+  using VecT = detail::change_base_type_t<T, uint8_t>;
+  auto DestP = multi_ptr<VecT, DestS>(reinterpret_cast<VecT *>(Dest.get()));
+  auto SrcP = multi_ptr<VecT, SrcS>(reinterpret_cast<VecT *>(Src.get()));
+  return async_group_copy(g, SrcP, DestP, NumElements, Stride);
+}
+
+/// Asynchronously copies a number of elements specified by \p numElements
+/// from the source pointed by \p src to destination pointed by \p dest and
+/// returns a SYCL device_event which can be used to wait on the completion
+/// of the copy.
+/// Permitted types for dataT are all scalar and vector types.
+template <typename Group, typename dataT>
+detail::enable_if_t<is_group_v<Group>, device_event>
+async_group_copy(Group g, global_ptr<dataT> src, local_ptr<dataT> dest,
+                 size_t numElements) {
+  return async_group_copy(g, src, dest, numElements, 1);
+}
+
+/// Asynchronously copies a number of elements specified by \p numElements
+/// from the source pointed by \p src to destination pointed by \p dest and
+/// returns a SYCL device_event which can be used to wait on the completion
+/// of the copy.
+/// Permitted types for dataT are all scalar and vector types.
+template <typename Group, typename dataT>
+device_event async_group_copy(Group g, local_ptr<dataT> src,
+                              global_ptr<dataT> dest, size_t numElements) {
+  return async_group_copy(g, src, dest, numElements, 1);
+}
+
+template <typename Group, typename... eventTN>
+void wait_for(Group g, eventTN... Events) {
+  // Events.wait() calls __spirv_GroupWaitEvents
+  // __spirv_GroupWaitEvents ignores event_list and calls __spirv_ControlBarrier
+  // https://github.com/intel/llvm/blob/sycl/libclc/generic/libspirv/async/wait_group_events.cl
+  // __spirv_ControlBarrier calls __syncthreads or __nvvm_bar_warp_sync
+  // https://github.com/intel/llvm/blob/sycl/libclc/ptx-nvidiacl/libspirv/synchronization/barrier.cl
+  (Events.ext_oneapi_wait(g), ...);
```

Comment:
Folding expressions are C++ 17 feature. @romanovvlad should we guard those with macros for extensions?

Reply:
It depends, there is a test: sycl/test/basic_tests/stdcpp_compat.cpp. If it passes there is no need to guard.

Reply:
Works fine on my machine